### PR TITLE
Replace amdgcn.ldexp with llvm.ldexp

### PR DESCRIPTION
**Pull request by Flakebi**
There is llvm.ldexp since https://reviews.llvm.org/D14327.
GlobalISel is failing some tests because it does not understand
amdgcn.ldexp anymore.

### CODE REVIEW

**Comment by amdvlk-admin** [comment]
173f5e10ee37b995018eb6fbea98d4f1c9af7586 Jenkins build error.
error: Leaf constant_fold_fabs is unreachable
note: Leaf idempotent_prop will have already matched
error: Leaf constant_fold_fabs is unreachable
note: Leaf idempotent_prop will have already matched
[1154/3070] Building AMDGPUGenSubtargetInfo.inc...
[1155/3070] Building AMDGPUGenPostLegalizeGICombiner.inc...
error: Leaf constant_fold_fabs is unreachable
note: Leaf idempotent_prop will have already matched
error: Leaf constant_fold_fabs is unreachable
note: Leaf idempotent_prop will have already matched
[1156/3070] Building CXX object compiler/llpc/llvm/lib/CodeGen/CMakeFiles/LLVMCodeGen.dir/CFIInstrInserter.cpp.o
[1157/3070] Building CXX object compiler/llpc/llvm/lib/CodeGen/CMakeFiles/LLVMCodeGen.dir/DeadMachineInstructionElim.cpp.o
[1158/3070] Building CXX object compiler/llpc/llvm/lib/CodeGen/CMakeFiles/LLVMCodeGen.dir/CodeGenCommonISel.cpp.o
[1159/3070] Building CXX object compiler/llpc/llvm/lib/CodeGen/CMakeFiles/LLVMCodeGen.dir/DetectDeadLanes.cpp.o
--
/jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llpc/lgc/builder/ArithBuilder.cpp:837:47: error: ‘ldexp’ is not a member of ‘llvm::Intrinsic’; did you mean ‘exp’?
  837 |     Value *ldexp = CreateIntrinsic(Intrinsic::ldexp, {x->getType(), exp->getType()}, {x, exp});
      |                                               ^~~~~
      |                                               exp
[2855/3070] Building CXX object compiler/llpc/llvm/tools/lgc/CMakeFiles/LLVMlgc.dir/builder/Builder.cpp.o
[2856/3070] Building CXX object compiler/llpc/CMakeFiles/llpc_standalone_compiler.dir/tool/llpcCompilationUtils.cpp.o

File: lgc/state/PalMetadata.cpp [hunk 1]
```diff
@@ -383,7 +383,7 @@ void PalMetadata::mergeFromBlob(llvm::StringRef blob, bool isGlueCode) {
       if (destNode->getKind() == msgpack::Type::UInt)
         *destNode = destNode->getUInt() | srcNode.getUInt();
       else if (destNode->getKind() == msgpack::Type::Boolean)
-        *destNode = destNode->getBool() | srcNode.getBool();
+        *destNode = destNode->getBool() || srcNode.getBool();
```
Comment (jayfoad) [inline lgc/state/PalMetadata.cpp:386]: This is unrelated, but I don't mind if you want to leave it in the patch.

**Review by jayfoad — DISMISSED** [review]

File: lgc/patch/PatchWaveSizeAdjust.cpp [hunk 2]
```diff
@@ -129,6 +129,7 @@ bool PatchWaveSizeAdjust::is16BitArithmeticOp(Instruction *inst) {
     case Intrinsic::umax:
     case Intrinsic::smax:
     case Intrinsic::fma:
+    case Intrinsic::ldexp:
```
[the remark this answers is not in the document]
Comment (Flakebi) [reply lgc/patch/PatchWaveSizeAdjust.cpp:132]: From the code, it seems like we use wave64 for every shader that contains a binary or unary operator (or any of these intrinsics).

Wave64 is not a bad default choice for gfx11, but that could be achieved faster than iterating over all instructions 🤔

[the remark this answers is not in the document]
Comment (jayfoad) [reply lgc/patch/PatchWaveSizeAdjust.cpp:132]: > every shader that contains a binary or unary operator

I'm guessing that's _most_ of them...?

**Comment by amdvlk-admin** [comment]
### Test summary for commit 2eef1e25406987b261df1be2c830632d25866046
<details><summary>CTS tests (Failed: 0/141320)</summary><ul><li>Built with version 1.3.3.1</li>
<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 40065/70667 (56.7%)</li>
<li>Failed: 0/70667 (0.0%)</li>
<li>Not Supported: 30602/70667 (43.3%)</li><li>Warnings: 0/70667 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Navi2x</summary><ul>

<li>Passed: 40106/70653 (56.8%)</li>
<li>Failed: 0/70653 (0.0%)</li>
<li>Not Supported: 30547/70653 (43.2%)</li><li>Warnings: 0/70653 (0.0%)</li>
</ul></details>


</ul></details>

File: lgc/patch/PatchWaveSizeAdjust.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (Flakebi) [reply lgc/patch/PatchWaveSizeAdjust.cpp:132]: I fixed that in #2508.

File: lgc/state/PalMetadata.cpp (hunk 1 above)
Comment (nhaehnle) [reply lgc/state/PalMetadata.cpp:386]: There's the exact same thing two lines above...

**Review by nhaehnle — APPROVED** [review]

**Pull request merged**